chalk-solve/src/clauses/builtin_traits/sized.rs

```diff
@@ -30,16 +20,66 @@ pub fn add_sized_program_clauses<I: Interner>(
         return;
     }
 
+    let interner = db.interner();
+
     // To check if a struct type S<..> is Sized, we only have to look at its last field.
     // This is because the WF checks for structs require that all the other fields must be Sized.
     let last_field_ty = struct_datum
         .binders
         .map_ref(|b| b.fields.last().unwrap())
         .substitute(interner, substitution);
 
-    let last_field_sized_goal = TraitRef {
-        trait_id: trait_ref.trait_id,
-        substitution: Substitution::from1(interner, last_field_ty),
-    };
-    builder.push_clause(trait_ref.clone(), iter::once(last_field_sized_goal));
+    needs_impl_for_tys(db, builder, trait_ref, iter::once(last_field_ty));
+}
+
+fn push_tuple_sized_conditions<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_ref: &TraitRef<I>,
+    arity: usize,
+    substitution: &Substitution<I>,
+) {
+    // Empty tuples are always Sized
+    if arity == 0 {
+        builder.push_fact(trait_ref.clone());
+        return;
+    }
+
+    let interner = db.interner();
+
+    // To check if a tuple is Sized, we only have to look at its last element.
+    // This is because the WF checks for tuples require that all the other elements must be Sized.
```

Comment:
👍 to comment!

Reply:
It seems like `push_tuple_sized_condition` and `push_struct_sized_conditions` could be coallesced into one function, though. Something that takes a type plus a list of "constituent types" and picks the last one from the list.

Reply:
For the comment, i copied the one for struct made by @Areredify . He is the one to be thanked for :+1:

Reply:
About coalescing both functions, i feel like `needs_impl_for_tys` is doing a good job here already. What would be the prototype of the function you have in mind?

Reply:
Yeah, ok, I take it back. I think there *is* a kernel of logic you could extract, but it's not worth it.